src/auth/mysql.rs

```diff
@@ -0,0 +1,362 @@
+/// a randomly generated salt.
+pub struct MySqlAuthenticator {
+    database_uri: String
+}
+
+static CHARS: &'static [u8] = b"0123456789abcdef";
+
+impl MySqlAuthenticator {
+    /// Create a new `MySqlAuthenticator` with a database connection
+    ///
+    pub fn new(uri: String) -> Self {
+        MySqlAuthenticator {database_uri: uri}
+    }
+
+    /// Create a new `MySqlAuthenticator` with a database config
+    ///
+    pub fn with_configuration(host: &str, port: u16, database: &str, user: &str, pass: &str) -> Result<Self, Error> {
+        let database_uri: String = String::from(
+            format!("mysql://{}:{}@{}:{}/{}", user, pass, host, port, database)
+        );
+        let authenticator = MySqlAuthenticator{database_uri};
+        match authenticator.test_connection() {
+            Ok(_) => Ok(authenticator),
+            Err(e) => Err(Error::GenericError(e.to_string())),
+        }
+    }
+
+    fn test_connection(&self) -> Result<&'static str, Error> {
+        let _ = self.connect()?;
+        Ok("Connection successful")
+    }
+
+    /// Connects to MySql Server
+    fn connect(&self) -> Result<MysqlConnection, Error> {
+        debug_!("Connecting to Mysql server");
+        let connection = MysqlConnection::establish(&self.database_uri)?;
+        Ok(connection)
+    }
+
+    /// Search for the specified user entry
+    fn search(
+        &self,
+        connection: &MysqlConnection,
+        search_user: &str
+    ) -> Result<Vec<User>, Error> {
+        use super::schema::users::dsl::*;
+        let results = users.filter(username.eq(search_user))
+            .load::<User>(connection)
+            .map_err(|e| Error::GenericError(
+                    format!("Database query failed: {}", e)
+            ))?;
+        Ok(results)
+    }
+
+    /// Hash a password with the salt. See struct level documentation for the algorithm used.
+    // TODO: Write an "example" tool to salt easily
+    pub fn hash_password(password: &str, salt: &[u8]) -> Result<String, Error> {
+        Ok(hex_dump(
+            Self::hash_password_digest(password, salt)?.as_ref(),
+        ))
+    }
+
+    /// Verify that some user with the provided password exists in the CSV database, and the password is correct.
+    /// Returns the payload to be included in a refresh token if successful
+    pub fn verify(
+        &self,
+        username: &str,
+        password: &str,
+        include_refresh_payload: bool,
+    ) -> Result<AuthenticationResult, Error> {
+        let user = {
+            let connection = self.connect()?;
+            let mut user = self.search(&connection, username).map_err(|_e| {
+                super::Error::AuthenticationFailure
+            })?;
+            if user.len() != 1 {
+                Err(super::Error::AuthenticationFailure)?;
+            }
+
+            user.pop().unwrap() // safe to unwrap
+        };
+        if username != user.username {
```

Comment:
IMO this should be an `assert_eq!` which should panic if this predicate is not met. This will likely only result if there is a bug in either Diesel or the query crafted in `search()`.